<a href="https://colab.research.google.com/github/KWEYUGEOFREY/Data-Science/blob/master/Cab_Project_Data_Analysis_SQL_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### a) Specifying the Data Analysis Question

Create a descriptive analysis report that outlines the pricing patterns of the existing ride-sharing company.

### b) Defining the Metric for Success

The project will be a success when we are able to understand how pricing works.

### c) Understanding the Context 

Over the past few years, ride-sharing apps have been on the rise across many cities in the world. While this has happened, Uber and Lyft's ride prices are not constant like public transport. They are greatly affected by the demand and supply of rides at a given time.

As a Data Scientist working to understand this market, you have been tasked to come up with a descriptive analysis report to help a Ride-Sharing Startup coming into this space, understand the various patterns on how pricing works for the existing ride-sharing company.

### d) Recording the Experimental Design

1. Load pandas and sqlite3
2. Load datasets 
3. Perform data cleaning
4. Transfer datasets to SQL table
5. Carry out analysis using SQL
6. Merge datasets and carry out further analysis
6. Summarize findings.
7. Provide recommendations.
8. Challenge the solution.

# Loading sql extensions and useful packages


In [ ]:
# Loading sql extensions
%load_ext sql
%sql sqlite:///testdb.sqlite   

'Connected: @testdb.sqlite'

In [ ]:
# import Python Pandas library
import pandas as pd
import numpy as np

# import sqlite3
import sqlite3


# Loading datasets 

In [ ]:
# load first few records
df1 = pd.read_csv('https://bit.ly/cabsdataset')

In [ ]:
# create cabs table

%sql DROP TABLE IF EXISTS df1;
%sql PERSIST df1;
%sql SELECT * FROM df1 limit 10;

 * sqlite:///testdb.sqlite
Done.
 * sqlite:///testdb.sqlite
 * sqlite:///testdb.sqlite
Done.


[(0, 0, 0.44, 'Lyft', '2018-12-16 09:30:07.890', 'North Station', 'Haymarket Square', 5.0, 1.0, '424553bb-7174-41ea-aeb4-fe06d4f4b9d7', 'lyft_line', 'Shared'),
 (1, 1, 0.44, 'Lyft', '2018-11-27 02:00:23.677', 'North Station', 'Haymarket Square', 11.0, 1.0, '4bd23055-6827-41c6-b23b-3c491f24e74d', 'lyft_premier', 'Lux'),
 (2, 2, 0.44, 'Lyft', '2018-11-28 01:00:22.198', 'North Station', 'Haymarket Square', 7.0, 1.0, '981a3613-77af-4620-a42a-0c0866077d1e', 'lyft', 'Lyft'),
 (3, 3, 0.44, 'Lyft', '2018-11-30 04:53:02.749', 'North Station', 'Haymarket Square', 26.0, 1.0, 'c2d88af2-d278-4bfd-a8d0-29ca77cc5512', 'lyft_luxsuv', 'Lux Black XL'),
 (4, 4, 0.44, 'Lyft', '2018-11-29 03:49:20.223', 'North Station', 'Haymarket Square', 9.0, 1.0, 'e0126e1f-8ca9-4f2e-82b3-50505a09db9a', 'lyft_plus', 'Lyft XL'),
 (5, 5, 0.44, 'Lyft', '2018-12-17 18:25:12.138', 'North Station', 'Haymarket Square', 16.5, 1.0, 'f6f6d7e4-3e18-4922-a5f5-181cdd3fa6f2', 'lyft_lux', 'Lux Black'),
 (6, 6, 1.08, 'Lyft', '2018-11-26 05:03:00.200', 'Northeastern University', 'Back Bay', 10.5, 1.0, '462816a3-820d-408b-8549-0b39e82f65ac', 'lyft_plus', 'Lyft XL'),
 (7, 7, 1.08, 'Lyft', '2018-12-02 19:53:04.677', 'Northeastern University', 'Back Bay', 16.5, 1.0, '474d6376-bc59-4ec9-bf57-4e6d6faeb165', 'lyft_lux', 'Lux Black'),
 (8, 8, 1.08, 'Lyft', '2018-12-03 06:28:02.645', 'Northeastern University', 'Back Bay', 3.0, 1.0, '4f9fee41-fde3-4767-bbf1-a00e108701fb', 'lyft_line', 'Shared'),
 (9, 9, 1.08, 'Lyft', '2018-11-27 10:45:22.249', 'Northeastern University', 'Back Bay', 27.5, 1.0, '8612d909-98b8-4454-a093-30bd48de0cb3', 'lyft_luxsuv', 'Lux Black XL')]

In [ ]:
# load first few records
df2 = pd.read_csv('https://bit.ly/cabsweatherdata')
df2.head(5)

Unnamed: 0   temp            location  ...           time_stamp  humidity   wind
0           0  42.42            Back Bay  ...  2018-12-16 23:45:01      0.77  11.25
1           1  42.43         Beacon Hill  ...  2018-12-16 23:45:01      0.76  11.32
2           2  42.50   Boston University  ...  2018-12-16 23:45:01      0.76  11.07
3           3  42.11              Fenway  ...  2018-12-16 23:45:01      0.77  11.09
4           4  43.13  Financial District  ...  2018-12-16 23:45:01      0.75  11.49

[5 rows x 9 columns]

In [ ]:
%sql DROP TABLE IF EXISTS  df2;
%sql PERSIST df2;
%sql SELECT * FROM df2 limit 5;

 * sqlite:///testdb.sqlite
Done.
 * sqlite:///testdb.sqlite
 * sqlite:///testdb.sqlite
Done.


[(0, 0, 42.42, 'Back Bay', 1.0, 1012.14, 0.1228, '2018-12-16 23:45:01', 0.77, 11.25),
 (1, 1, 42.43, 'Beacon Hill', 1.0, 1012.15, 0.1846, '2018-12-16 23:45:01', 0.76, 11.32),
 (2, 2, 42.5, 'Boston University', 1.0, 1012.15, 0.1089, '2018-12-16 23:45:01', 0.76, 11.07),
 (3, 3, 42.11, 'Fenway', 1.0, 1012.13, 0.0969, '2018-12-16 23:45:01', 0.77, 11.09),
 (4, 4, 43.13, 'Financial District', 1.0, 1012.14, 0.1786, '2018-12-16 23:45:01', 0.75, 11.49)]

#   

# Reformating

In [ ]:

%%sql 
select location, round(rain, 2) as rainfall
from df2 limit 5;

 * sqlite:///testdb.sqlite
Done.


[('Back Bay', 0.12),
 ('Beacon Hill', 0.18),
 ('Boston University', 0.11),
 ('Fenway', 0.1),
 ('Financial District', 0.18)]

In [ ]:
#getting rid of microseconds in the time_stamp column
df1['time_stamp'] = df1['time_stamp'].astype(str).str[:-4].astype(str)


In [ ]:
# connecting to the database and transferring the data from the Pandas dataframes to the SQL tables

conn = sqlite3.connect('testdb.sqlite')
df1.to_sql('df1', conn, index = False, if_exists = "replace")
df2.to_sql('df2', conn, index = False, if_exists = "replace")

/usr/local/lib/python3.6/dist-packages/pandas/core/generic.py:2663: UserWarning: The spaces in these column names will not be changed. In pandas versions < 0.14, spaces were converted to underscores.
  method=method,


In [ ]:
#joining the tables

%%sql

DROP TABLE if EXISTS df3;

CREATE TABLE df3 as
SELECT df1.distance,	df1.cab_type,	df1.time_stamp,	df1.destination,	
df1.source,	df1.price,	df1.surge_multiplier,	df1.id,	df1.product_id,	
df1.name, df2.temp,	df2.location,	df2.clouds,	df2.pressure,	
df2.rain,	df2.humidity,	df2.wind
from df1
INNER JOIN df2 ON
df1.time_stamp = df1.time_stamp;

SELECT * from df3 LIMIT 5;

 * sqlite:///testdb.sqlite
Done.
(sqlite3.OperationalError) database or disk is full
[SQL: CREATE TABLE df3 as
SELECT df1.distance,	df1.cab_type,	df1.time_stamp,	df1.destination,	
df1.source,	df1.price,	df1.surge_multiplier,	df1.id,	df1.product_id,	
df1.name, df2.temp,	df2.location,	df2.clouds,	df2.pressure,	
df2.rain,	df2.humidity,	df2.wind
from df1
INNER JOIN df2 ON
df1.time_stamp = df1.time_stamp;]
(Background on this error at: http://sqlalche.me/e/13/e3q8)


#Questions to answer from our data

In [ ]:
#checking pricing agains locations
%%sql 

select location, cab_type, sum(price) as "Standard price" 
from df3
group by location
order by "Standard price" desc
limit 50;


 * sqlite:///testdb.sqlite
(sqlite3.OperationalError) no such table: df3
[SQL: select location, cab_type, sum(price) as "Standard price" 
from df3
group by location
order by "Standard price" desc
limit 50;]
(Background on this error at: http://sqlalche.me/e/13/e3q8)


In [ ]:
#Checking for what time and weather conditions associates with surge in price
%%sql

select cab_type, time_stamp, wind, temp,  rain, surge_multiplier
from df3
where not rain ="None"
order by surge_multiplier desc
limit 10

UsageError: Cell magic `%%sql` not found.


In [ ]:
#Checking for what time uber charges surge
%%sql
select cab_type, time_stamp, surge_multiplier, temp
from df3
where cab_type = "Uber"
order by surge_multiplier desc
limit 10;

In [ ]:
#Checking for what time lyft charges surge
%%sql
select cab_type, time_stamp, surge_multiplier, temp
from df3
where cab_type = "Lyft"
order by surge_multiplier desc
limit 10;

In [ ]:
#Checking for what time fare charges surge
%%sql
select name, count(name) as "Number of orders", round(avg(price),2) as "Standard price", cab_type, time_stamp
from df3
group by name
order by "Standard price" desc
limit 10;

In [ ]:
##Checking fare charges surge on hot days
%%sql

select avg(price) as "Standard price", cab_type
from df3
where temp >= 30.0
group by cab_type;

limit 10;

In [ ]:
#total amount in the morning hours
%%sql

select sum(price) as "Total_amount_morning"
from df3
where time_stamp between '2018-11-26 1:00:00' and '2018-11-26 9:59:59'

In [ ]:
#during the day prices
%%sql

select sum(price) as "Total_amount_daytime"
from df3
where time_stamp between '2018-11-26 10:00:00' and '2018-11-26 15:59:59'

In [ ]:
#prices at night
%%sql

select sum(price) as "Total_amount_night"
from 
where time_stamp between '2018-11-26 05:00:00' and '2018-11-26 11:59:59'

# Final thoughts

The following are the prices for Lyft:
- overall average of 17.35, overall minimum 2.5, overall maximum 97.5
- average 17.3, minimum 2.5, maximum 92.0 from midnight to 6 AM
- average 16.5, minimum 2.5, maximum 89.0 from 7 AM to 12 PM
- average 17.4, minimum 2.5, maximum 92.0 from 1 PM to 6 PM
- average 18.3, minimum 2.5, maximum 97.5 from 7 PM to 11 PM

The following are the prices for Uber:
- overall average of 15.8, overall minimum 4.5, overall maximum 89.5
- average 15.9, minimum 4.5, maximum 89.5 from midnight to 7 AM
- average 15.2, minimum 4.5, maximum 87.0 from 7 AM to 12 PM
- average 15.9, minimum 4.5, maximum 62.5 from 1 PM to 6 PM
- average 16.3, minimum 4.5, maximum 76.0 from 7 PM to 11 PM

Minimum distance is 0.02, maximum distance is 7.86.

**For distance 0.00-1.96:**
Lyft has average price 14.1, minimum price 2.5, and maximum price 65.0.
Uber has average price 13.3, minimum price 4.5, and maximum price 61.5.

**For distance 1.97-3.93:**
Lyft has average price 19.2, minimum price 2.5, and maximum price 92.0.
Uber has average price 17.1, minimum price 6.0, and maximum price 80.5.

**For distance 3.94-5.89:**
Lyft has average price 25.5,  minimum price 3.0, and maximum price 97.5.
Uber has average price 22.3, minimum price 7.5, and maximum price 87.0.

**For distance 5.90-7.86**
Lyft has average price 30.1, minimum price 10.5, and maximum price 65.0.
Uber has average price 26.1, minimum price 10.0, and maximum price 89.5.

**From the observation made, time of the day and weather highly contribute to pricing of cabs:**

**With bad weather prices are likely to shoot.**

**Also prices are high during morning and evening hours**

**Another factor affecting pricing is location of your ride**
